inc/classes/Validator/Email.php

```diff
@@ -0,0 +1,160 @@
+
+    /**
+     * Error type "format"
+     *
+     * When an email has a wrong format (e.g. @ is missing)
+     */
+    const INVALID_FORMAT_ERROR = 2;
+
+    /**
+     * Error type "mx record"
+     *
+     * When the host doesn't have a MX DNS record entered
+     */
+    const MX_CHECK_FAILED_ERROR = 3;
+
+    /**
+     * Error type "host record"
+     *
+     * When the host doesn't have a MX, A or AAAA DNS record entered
+     */
+    const HOST_CHECK_FAILED_ERROR = 4;
+
+    /**
+     * Option "MX record"
+     *
+     * If enabled, the MX DNS record will be checked for the host
+     */
+    const OPTION_MX_CHECK = 'mx-record-check';
+
+    /**
+     * Option "Host record"
+     *
+     * If enabled, the MX, A or AAAA DNS record will be checked for the host
+     */
+    const OPTION_HOST_CHECK = 'host-record-check';
+
+    /**
+     * Stores the validation mode.
+     * See VALIDATION_MODE_* constants.
+     *
+     * @var string
+     */
+    private $mode;
+
+    /**
+     * @param string $mode
+     */
+    public function __construct($mode = self::VALIDATION_MODE_HTML5)
+    {
+        $this->mode = $mode;
+
+        $this->options = [
+            self::OPTION_MX_CHECK => false,
+            self::OPTION_HOST_CHECK => false,
+        ];
+    }
+
+    /**
+     * Checks if the passed value is valid.
+     *
+     * @param string      $value      The value that should be validated
+     * @return boolean
+     */
+    public function validate($value)
+    {
+        if (null === $value || '' === $value) {
+            $this->errorCode = self::EMAIL_EMPTY_ERROR;
+            return false;
+        }
+
+        $value = (string) $value;
+
+        if (!preg_match(self::$emailPatterns[$this->mode], $value)) {
+            $this->errorCode = self::INVALID_FORMAT_ERROR;
+            return false;
+        }
+
+        $host = (string) substr($value, strrpos($value, '@') + 1);
+
+        // Check for host DNS resource records
+        if ($this->isOptionEnabled(self::OPTION_MX_CHECK)) {
+            if (!$this->checkMX($host)) {
+                $this->errorCode = self::MX_CHECK_FAILED_ERROR;
+                return false;
+            }
+
+            return true;
+        }
+
+        if ($this->isOptionEnabled(self::OPTION_HOST_CHECK) && !$this->checkHost($host)) {
+            $this->errorCode = self::HOST_CHECK_FAILED_ERROR;
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check DNS Records for MX type.
+     *
+     * @param string $host
+     * @return bool
+     */
+    private function checkMX(string $host): bool
+    {
+        return '' !== $host && checkdnsrr($host, 'MX');
+    }
+
+    /**
+     * Check if one of MX, A or AAAA DNS RR exists.
+     *
+     * @param string $host
+     * @return bool
+     */
+    private function checkHost(string $host): bool
+    {
+        return '' !== $host && ($this->checkMX($host) || (checkdnsrr($host, 'A') || checkdnsrr($host, 'AAAA')));
```

Comment:
I suspect that this would completely disable registration for any system that cannot reach a DNS resolver (misconfiguration, connectivity issues or local installation without internet access) as the return value FALSE  of `checkdnsrr` is set for both technical issues and missing records.
Would it make sense to use [`dns-get-record()`](http://php.net/manual/de/function.dns-get-record.php) instead?

Reply:
Ahhh, you got me there!
Misread the documentation, I thought it would also return something as array if the lookup fails.
So no benefit from using a different built-in function.
There are pure PHP implementations (e.g. [NetDNS2](https://github.com/mikepultz/netdns2)) that throw useful exceptions, but I suspect that would be over-engineering a simple problem.

I would argue against the two configuration settings though.
Wouldn't it make more sense to put these as setting into of the usermgmt-module or general system config? (maybe even as general "enable/disable DNS lookups"?)

We could also simply define that the DNS verification only blocks registration on internet-systems and only shows a warning for intranet-systems. 
DNS configuration should then be checked during the installation.

What do you think?

Reply:
> but I suspect that would be over-engineering a simple problem
I agree here.

> Wouldn't it make more sense to put these as setting into of the usermgmt-module or general system config? (maybe even as general "enable/disable DNS lookups"?)
I agree. From the first glance, this would make sense. Let me have a look later this day how we can implement this.

> We could also simply define that the DNS verification only blocks registration on internet-systems and only shows a warning for intranet-systems.
True in general. But i would split this feature and not include it here in this PR.
Why? Because i think we have multiple places that don't work in an intranet env (e.g. mails).
How is your view here?

Reply:
>True in general. But i would split this feature and not include it here in this PR.
Why? Because i think we have multiple places that don't work in an intranet env (e.g. mails).
How is your view here?

Simple: There is already an option to define if an installation is for internet/intranet (`$cfg["sys_internet"]`) which is used to switch functionality in a lot of places. Thus that does luckily not require any additional work